Add duplication capability to functions and outputs

internalthought opened the pull request:
Inspired by issue #1366

kizniche commented:
I see you found my clone_model function I used to duplicate desktops. This looks great. I'll test soon to see if I notice any issues. Thanks for the PR!

kizniche commented:
The test test_add_all_function_devices_logged_in_as_admin seems to be failing.

 =========================== short test summary info ============================
FAILED tests/software_tests/test_mycodo_flask/test_endpoints.py::test_add_all_function_devices_logged_in_as_admin - AssertionError: Number of Functions doesn't match: In DB 1, Should be: 0
assert 1 == 0
FAILED tests/software_tests/test_mycodo_flask/test_endpoints.py::test_add_all_function_actions_logged_in_as_admin - AssertionError: Number of Actions doesn't match: In DB 0, Should be: 1
assert 0 == 1

I'll investigate later

internalthought commented:
fixed my implementation to work with the clone_model -> database interactions. also had issues installing, so put in some workarounds to get it running. if its too messy, i can clean that up.

kizniche commented:
What is the purpose of the additional commands in the update-permissions section of upgrade_commands.sh? Restarting all services when permissions are set seems out of place and is likely to negatively impact some users.

internalthought commented:
for some reason, when running setup.sh, the web server connect test doesn't pass unless i delete the mycodoflask.sock and restart all three services. i reordered the setup.sh commands and added to the reset web server command, just to pass through install. not entirely sure why that's the case. i fixed the endpoint test to work with duplication test case, although my changes may make this test less useful. either way, the duplication function and buttons work and pytest passes!

kizniche commented on mycodo/scripts/upgrade_commands.sh:
```diff
@@ -686,11 +701,29 @@ case "${1:-''}" in
+        printf "#### Ensuring socket directory permissions\n"
+        mkdir -p /usr/local
+        chmod 775 /usr/local
+        chown root:mycodo /usr/local
```
We shouldn't be changing the group of a user's /usr/local directory
